csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,457 @@
+// return true is that `binaryOp(0, x) == 0` & `binaryOp(x, 0) == 0`
+bool padCompatibleBinaryOp(BinaryOpType t) {
+  switch (t) {
+    case BinaryOpType::Add:
+    case BinaryOpType::Mul:
+    case BinaryOpType::Sub:
+    case BinaryOpType::BitwiseAnd:
+    case BinaryOpType::LogicalAnd:
+      return true;
+    default:
+      return false;
+  }
+}
+
+// This function checks if we could propagate the PadOp on `val` to its
+// definition. i.e. changing the graph from
+//   inp --> definition() --> val --> PadOp --> padded_val
+// to
+//   inp --> PadOp --> padded_inp --> definition() --> padded_val
+//
+// returns true if `PadOp` should be propagated pass val->definition().
+//
+// NOTE 0: For details on the propagation logic, see Note [ PadOp Propagation
+// Rule ] NOTE 1: This function also update frontier accordingly, see note
+// below.
+bool shouldPropagatePad(Val* val, std::unordered_map<Val*, int64_t>& frontier) {
+  // NOTE [ Handling TV with Multiple Uses via Frontier ]
+  //
+  // Frontier counts the encounter of each Val during the propagation. When we
+  // reach a Val during the propagation, we add Val entry *temporarily* to the
+  // frontier map and increment its encounter count. We cannot propagate the
+  // PadOp over the producer of Val yet, because it still have other uses that
+  // expects the original Val. Until we see all uses of Val during the
+  // traversal, we know all uses can be replaced with the padded Val and hence
+  // it's safe to propagate `PadOp` across. We remove Val fron `frontier` map
+  // since we can push the replay to its producer. We return `true` to signal
+  // that further propagation is allowed.
+  if (val->isFusionOutput()) {
+    // skipping propagation when target is an output.
+    frontier.emplace(val, -1);
+    return false;
+  }
+
+  int64_t current_use = frontier.count(val) == 0 ? 1 : frontier[val] + 1;
+
+  // `current_use == 0` means `previous use == -1`, blocking propagation
+  if (current_use == 0) {
+    return false;
+  }
+
+  if (current_use == static_cast<int64_t>(val->uses().size())) {
+    // all uses of the entry has been encounter in this traversal, we can
+    // safely propagate it across. remove val from frontier map
+    frontier.erase(val);
+    return true;
+  }
+
+  // updating uses in frontier. return false since we are not yet ready to
+  // propagate
+  frontier[val] = current_use;
+  return false;
+}
+
+// replayConcretePad tries to replay a concretized PadOp on pad_tv. This
+// function allows merging multiple consecutive PadOp by stacking their pad
+// widths together as `vec_pad_widths`. This function assumes it has already
+// resolved the output iter_type for each IterDomain and provided as
+// `ref_iter_type`. The function returns the output from the replay PadOp.
+//
+// NOTE: this assumes all vec_pad_widths are positive entries so we don't need
+// to consider accumulating them changing the output iter_type.
+TensorView* replayConcretePad(
```

Comment:
Can this be merged into https://github.com/NVIDIA/Fuser/blob/1bd98c19390c1bd3c2e54954b85315f905c0db2f/csrc/transform_replay.h#L296?

Reply:
I'm likely not following the idea here.

I can try to squeeze it inside that as an overload?! I'm not sure this function I have here is a general `replay expression` and feels that it's very specific just how we construct `pad`. Would you mind elaborating on your suggestion? :bow:

Reply:
Let me rephrase: replayExprWithNewInput was built to replay a unary op on a new input. It does so by replaying the IterDomain transforms on the new root domain. What's special about PadOp that would break this utility?

Reply:
> replay a unary op on a new input

I think it's the **replay** in this function name that's confusing :)

The difference here is that `PadOp` is not a unary op. i.e. in this utility function we do expect new `pad_value` as well as stacked pad widths. So we cannot just replay the old transformation on the exprssion.